### DISCUSSION TITLE
hash navigation hangs between page loads.

### DISCUSSION THREAD

**Discussion by bewutze** [h2oai/wave#1588] [Q&A] [answered]
When using the hash navigation with 0.22.0, the initial page loads.  Then when you select any other page there is a pending/load unless you refresh the page and then the one you selected loads.  How can I get this to load automatically?

In init:
`q.page['header'] = ui.header_card(
        box='header',
        title='Dashboard',
        subtitle="Daanesh Ibrahim",
        nav=[
            ui.nav_group('Menu', items=[
                ui.nav_item(name='#menu/..', label='..'),
                ui.nav_item(name='#menu/..', label='..'),
                ui.nav_item(name='#menu/..', label='..'),
            ])]
    )`

in home:
`hash = q.args['#']
if hash == '..':
   Do code`

**Comment by mturoci** [comment, marked as answer]
Hi @bewutze, the loading spinner appears when your Wave app doesn't update the UI in reaction to a user event, e.g. clicking the nav item. The problem in your case is probably that code in your if statement doesn't get executed. However, it's hard to guess without a min repro code. If you can provide one, I can give you the solution.

Meanwhile I also suggest reading the [docs](https://wave.h2o.ai/docs/routing/).

**Reply by bewutze** [reply]
Here is the minimal reproducing code:
Thanks, brains have been breaking on this one.
[app.zip](https://github.com/h2oai/wave/files/9353489/app.zip)
.

**Reply by bewutze** [reply]
run with

wave run app.py

**Reply by mtanco** [reply]
Thank you for sharing an example! There is nothing in the `serve` function that is telling wave what to do when a tab is clicked and the hash changes. Essentially, `await handle_on(q)` doesn't know that you want the `home` function to run when the hash changes. 

@mturoci may have another suggestion, but you can update your `serve` function to:

```python
async def serve(q: Q):
    # First time a browser comes to the app
    if not q.client.initialized:
        await init(q)
        q.client.initialized = True

    # Other browser interactions
    if not await handle_on(q):
        await home(q)
    await q.page.save()
```

A couple other things: 
* In the logs I am seeing error messages: ` RuntimeWarning: coroutine 'AsyncPage.save' was never awaited`. The `q.page.save()` function is asynchronous so that it can handle multi-users and requests, you must always include the `await` keyword like `await q.page.save()`. 
* You are already saving the page in the `serve` function so you can actually take this out of the if statements in the `home` function
* You are calling the `clear_cards` function multiple times, you can just do this at the beginning of `home`

Happy wave-ing!

**Reply by daaneshibrahim** [reply]
> Thank you for sharing an example! There is nothing in the `serve` function that is telling wave what to do when a tab is clicked and the hash changes. Essentially, `await handle_on(q)` doesn't know that you want the `home` function to run when the hash changes.
> 
> @mturoci may have another suggestion, but you can update your `serve` function to:
> 
> ```python
> async def serve(q: Q):
>     # First time a browser comes to the app
>     if not q.client.initialized:
>         await init(q)
>         q.client.initialized = True
> 
>     # Other browser interactions
>     if not await handle_on(q):
>         await home(q)
>     await q.page.save()
> ```
> 
> A couple other things:
> 
> * In the logs I am seeing error messages: ` RuntimeWarning: coroutine 'AsyncPage.save' was never awaited`. The `q.page.save()` function is asynchronous so that it can handle multi-users and requests, you must always include the `await` keyword like `await q.page.save()`.
> * You are already saving the page in the `serve` function so you can actually take this out of the if statements in the `home` function
> * You are calling the `clear_cards` function multiple times, you can just do this at the beginning of `home`
> 
> Happy wave-ing!

Thank you so much for your time figuring this out! We were really stumped but this fixed it perfectly. 

Thank you again!

**Reply by bewutze** [reply]
That did the trick!  Thank you so much for the solution and the very clear explanation.  Cheers!
